docs/table/masking.rst

```diff
@@ -197,3 +210,100 @@ attribute.
+|MaskedColumn| columns.
+
+It is important to recognize that the ``masked`` attribute for a table does not imply
+whether any of the column data are actually masked.  A table can have ``masked=True`` but
+not have any masked elements in any table column.  Starting with version 4.0 there are two
+table properties which give more useful information about masking:
+
+- ``has_masked_columns`` : table has at least one |MaskedColumn| column.  This does *not*
+  check if any data values are actually masked.
+- ``has_masked_values`` : table has one or more column data values which are masked.
+  This may be relatively slow for large tables as it requires checking the mask
+  values of each column.
+
+Starting with version 4.0 the term "masked table" should be reserved for the narrow and
+less-common case of a table created with ``masked=True``.  In most cases there should be
+no need worry about "masked" or "unmasked" at the table level, but instead focus on the
+individual columns.
+
+Auto-upgrade to masked
+----------------------
+
+Prior to version 4.0, adding a |MaskedColumn| or a new row with masked elements to a table
+with ``masked=False`` would set ``masked=True`` and automatically "upgrade" other
+columns to be masked.  In many cases this upgrade of the other columns was unnecessary and
+an annoyance.
+
+Starting with 4.0, new columns are added using the column type which is appropriate for
+the data.  For instance, if a numpy masked array is added, then that will turn into a
+|MaskedColumn|, but no other columns will be affected and the ``masked`` attribute will
+remain as ``False``.
+
+A commonly-encountered implication of this change is that tables read with
+`~astropy.table.Table.read` will *always* have ``masked=False``, and only columns with
+masked values will be |MaskedColumn|.  Prior to 4.0 if the input table had any masked
+values then the returned table would have ``masked=True`` and all |MaskedColumn| columns.
+An example is in the next section.
+
+Recovering the pre-4.0 behavior
+-------------------------------
+
+For code that requires every existing or newly-added column to be masked, it is now
+required to explicitly specify ``masked=True`` when creating the table.  Previously the
+table would be auto-upgraded to use |MaskedColumn| for all columns as soon as the first
+masked column was added.  If the table already exists (e.g. after using
+`~astropy.table.Table.read` to read a data file), then one needs to make a new table:
+
+.. doctest-skip::
+
+  >> dat = Table.read('data.fits')
+  >> dat = Table(dat, masked=True, copy=False)  # Convert to masked table
+  >> dat['new_column'] = [1, 2, 3, 4, 5]  # Will be added as a MaskedColumn
+
+For most applications this should not be necessary, and the preferred idiom is the more
+explicit version below:
+
+.. doctest-skip::
+
+  >> dat = Table.read('data.fits')
+  >> dat['new_column'] = MaskedColumn([1, 2, 3, 4, 5])
```

Comment:
And what about giving ``dat['new_column'] = np.ma.array([1, 2, 3, 4, 5])`` as an alternative? (just to avoid the ``MaskedColumn`` import)

Reply:
:+1:, but maybe `np.ma.MaskedArray` instead

Reply:
Done.